### PR TITLE
Ice refactoring

### PR DESCRIPTION
**Pull request by Saintis**
Refactors `cards-ice` structure.

Please review :smile: 

Each commit should deal with one change as described, please let me know if anything seems unnecessary or less readable than before.

Does two major changes:
- Refactors out commonly used subroutines, like "end the run" and traces.
- Extracts stuff common to "sets" of ice (like the constellation "kicker" ice).

### CODE REVIEW

File: src/clj/game/cards-ice.clj [hunk 1]
```diff
@@ -235,57 +388,44 @@
...
-                 :msg (msg "reveal "
-                           (join ", " (map #(str (:title %) " ("
-                                                 (:label (first (:abilities (card-def %)))) ")") targets)))}
-                {:label "Resolve a Grail ICE subroutine from HQ"
-                 :choices {:req #(and (:side % "Corp")
-                                      (in-hand? %)
-                                      (has-subtype? % "Grail"))}
-                 :effect (req (doseq [ice targets]
-                                (resolve-ability state side (first (:abilities (card-def ice))) card nil)))}]}
+   (grail-ice end-the-run)
```
Comment (zaroth) [inline src/clj/game/cards-ice.clj]: That's beautiful.

**Comment by nealterrell** [comment]
Looking good now.

**Comment by mtgred** [comment]
There are merge conflicts. Can you have a look?

**Comment by Saintis** [comment]
Fixed.

**Comment by mtgred** [comment]
:+1:

**Pull request merged**